components/codec/number.rs

```diff
@@ -0,0 +1,2096 @@
+    #[inline]
+    pub fn encode_u64_le(buf: &mut [u8], v: u64) {
+        LittleEndian::write_u64(buf, v)
+    }
+
+    /// Decodes an unsigned 64 bit integer from `buf` in little endian,
+    /// which is previously encoded via `encode_u64_le`.
+    ///
+    /// # Panics
+    ///
+    /// Panics when `buf.len() < 8`.
+    #[inline]
+    pub fn decode_u64_le(buf: &[u8]) -> u64 {
+        LittleEndian::read_u64(buf)
+    }
+
+    /// Encodes a signed 64 bit integer `v` to `buf` in little endian,
+    /// which is not memory-comparable.
+    ///
+    /// # Panics
+    ///
+    /// Panics when `buf.len() < 8`.
+    #[inline]
+    pub fn encode_i64_le(buf: &mut [u8], v: i64) {
+        LittleEndian::write_i64(buf, v)
+    }
+
+    /// Decodes a signed 64 bit integer from `buf` in little endian,
+    /// which is previously encoded via `encode_i64_le`.
+    ///
+    /// # Panics
+    ///
+    /// Panics when `buf.len() < 8`.
+    #[inline]
+    pub fn decode_i64_le(buf: &[u8]) -> i64 {
+        LittleEndian::read_i64(buf)
+    }
+
+    /// Encodes a 64 bit float number `v` to `buf` in little endian,
+    /// which is not memory-comparable.
+    ///
+    /// # Panics
+    ///
+    /// Panics when `buf.len() < 8`.
+    #[inline]
+    pub fn encode_f64_le(buf: &mut [u8], v: f64) {
+        LittleEndian::write_f64(buf, v)
+    }
+
+    /// Decodes a 64 bit float number from `buf` in little endian,
+    /// which is previously encoded via `encode_f64_le`.
+    ///
+    /// # Panics
+    ///
+    /// Panics when `buf.len() < 8`.
+    #[inline]
+    pub fn decode_f64_le(buf: &[u8]) -> f64 {
+        LittleEndian::read_f64(buf)
+    }
+
+    /// Encodes an unsigned 64 bit integer `v` to `buf` in VarInt encoding,
+    /// which is not memory-comparable. Returns thenumber of bytes that encoded.
+    ///
+    /// Note: VarInt encoding is slow, try avoid using it.
+    ///
+    /// # Panics
+    ///
+    /// Panics when `buf.len() < 10`.
+    pub fn encode_var_u64(buf: &mut [u8], mut v: u64) -> usize {
+        assert!(buf.len() >= MAX_VARINT64_LENGTH);
+
+        unsafe {
+            let mut ptr = buf.as_mut_ptr();
+
+            while v >= 0x80 {
+                // We have already checked buffer size at the beginning so that it is safe to
+                // directly access the buffer.
+                *ptr = 0x80 | (v & 0x7f) as u8;
+                ptr = ptr.offset(1);
+                v >>= 7;
+            }
+            *ptr = v as u8;
+            ptr.offset_from(buf.as_ptr()) as usize + 1
+        }
+    }
+
+    /// Decodes an unsigned 64 bit integer from `buf` in VarInt encoding.
+    /// Returns decoded result and the number of bytes that successfully decoded.
+    ///
+    /// This function is more efficient when `buf.len() >= 10`.
+    ///
+    /// # Errors
+    ///
+    /// Returns `Error::BufferTooSmall` if there is not enough space to decode the whole VarInt.
+    pub fn try_decode_var_u64(buf: &[u8]) -> Result<(u64, usize)> {
```

Comment:
The main difference between it and tikv version is whether it's optimized for small numbers between 0~128, which is commonly used in a real life world.

Reply:
I adjusted some order and now their performance is same for small values.